clang/lib/Sema/SemaDecl.cpp

```diff
@@ -7402,10 +7402,17 @@ NamedDecl *Sema::ActOnVariableDeclarator(
 
   // Static variables declared inside SYCL device code must be const or
   // constexpr
-  if (getLangOpts().SYCLIsDevice)
-    if (SCSpec == DeclSpec::SCS_static && !R.isConstant(Context))
+  if (getLangOpts().SYCLIsDevice) {
+    if (SCSpec != DeclSpec::SCS_static && !NewVD->hasGlobalStorage() &&
+        isDecoratedWithSyclAttribute<SYCLDeviceGlobalAttr>(NewVD->getType()))
+      Diag(D.getIdentifierLoc(), diag::err_sycl_device_global_incorrect_scope);
+
+    if (SCSpec == DeclSpec::SCS_static && !R.isConstant(Context) &&
+        !isDecoratedWithSyclAttribute<SYCLGlobalVariableAllowedAttr>(
+            NewVD->getType()))
```

Comment:
We can actually not do that. `SYCLDiagIfDeviceCode` emits the error only for SYCL device code scope which is always function scope. It is not valid to declare a `device_global` inside a function anyway. 
```suggestion
    if (SCSpec == DeclSpec::SCS_static && !R.isConstant(Context))
```
Of course the error message will be not quite right, but it will prevent users from declaring a non-const `device_global` in function scope (which is incorrect) while we don't have full diagnostics support.
But I'm not going to push on that, since either path is not ideal in my opinion and I'll leave it up to you and other reviewers.

Reply:
I see your point. I am tempted to leave it as it is and address it in the next PR if the reviewers are ok with it.

Reply:
I'm ok with addressing this in upcoming diagnostics PR since we don't really have correct diagnostic support anyway.